src/anomalib/data/kolektor.py

```diff
@@ -0,0 +1,315 @@
+"""Kolektor Surface-Defect Dataset (CC BY-NC-SA 4.0).
+
+Description:
+    This script contains PyTorch Dataset, Dataloader and PyTorch
+        Lightning DataModule for the Kolektor Surface-Defect dataset.
+    The dataset can be found at https://www.vicos.si/resources/kolektorsdd/
+
+License:
+    Kolektor Surface-Defect dataset is released under the Creative Commons
+    Attribution-NonCommercial-ShareAlike 4.0 International License
+    (CC BY-NC-SA 4.0)(https://creativecommons.org/licenses/by-nc-sa/4.0/).
+
+Reference:
+    - Tabernik, Domen, Samo Šela, Jure Skvarč, and Danijel Skočaj.
+    "Segmentation-based deep-learning approach for surface-defect detection."
+    Journal of Intelligent Manufacturing 31, no. 3 (2020): 759-776.
+"""
+
+
+from __future__ import annotations
+
+import logging
+from pathlib import Path
+from urllib.request import urlretrieve
+from zipfile import ZipFile
+
+import albumentations as A
+import numpy as np
+from imageio.v2 import imread
+from pandas import DataFrame
+from sklearn.model_selection import train_test_split
+
+from anomalib.data.base import AnomalibDataModule, AnomalibDataset
+from anomalib.data.task_type import TaskType
+from anomalib.data.utils import (
+    InputNormalizationMethod,
+    Split,
+    TestSplitMode,
+    ValSplitMode,
+    get_transforms,
+)
+from anomalib.data.utils.download import hash_check
+
+logger = logging.getLogger(__name__)
+
+
+# Check if a mask shows defects
+def is_good(path):
+    img_arr = imread(path)
+    if np.all(img_arr == 0):
+        return 1
+    return 0
+
+
+def download_and_extract_kolektor(url: str, root: Path, file_hash: str):
+    """Download and extract a dataset.
+
+    Args:
+        url  (Str): Url for directly downloading the dataset
+        root (Path): Root directory where the dataset will be stored.
+        file_hash (Str): MD5 Checksum of the correct zip file
+    """
+    root.mkdir(parents=True, exist_ok=True)
+
+    downloaded_file_path = Path.joinpath(root, "KolektorSDD.zip")
+    if downloaded_file_path.exists():
+        logger.info("Existing dataset archive found. Skipping download stage.")
+    else:
+        logger.info("Downloading the Kolektor dataset archive")
+        urlretrieve(  # nosec - suppress bandit warning (urls are hardcoded)
+            url=url,
+            filename=downloaded_file_path,
+        )
+        logger.info("Checking the hash of the downloaded file.")
+        hash_check(downloaded_file_path, file_hash)
+
+    logger.info("Extracting dataset into given folder")
+    with ZipFile(downloaded_file_path, "r") as zip_file:
+        zip_file.extractall(root)
+
+    logger.info("Cleaning up files.")
+    (downloaded_file_path).unlink()
```

Comment:
There is a function to download and extract data [here](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/data/utils/download.py#L220). Is there a specific difference between your implementation and this implementation? If not, this would be a duplication.

Reply:
There's a slight difference as I've mentioned [here](https://github.com/openvinotoolkit/anomalib/pull/983#issuecomment-1498479796) The `download_and_extract` function in `utils` downloads the Kolektor dataset without an extension and therefore it can't be extracted. This happens since the provided link for the dataset doesn't specify the extension which is required by the `download_and_extract` function. In this implementation I've hardcoded the .zip extension in line 65.

Reply:
Maybe we could update the existing code a little bit. How about the following:

Replace
https://github.com/openvinotoolkit/anomalib/blob/d81dbb5b3dfd831576d137fb93881b34254a7fff/src/anomalib/data/utils/download.py#L25-L31

with:
```python
@dataclass
class DownloadInfo:
    """Info needed to download a dataset from a url."""

    name: str
    url: str
    hash: str
    filename: str | None = None
```

And replace
https://github.com/openvinotoolkit/anomalib/blob/d81dbb5b3dfd831576d137fb93881b34254a7fff/src/anomalib/data/utils/download.py#L229-L230

with,
```python
    # save the compressed file in the specified root directory, using the same file name as on the server
    if info.filename:
        downloaded_file_path = root / info.filename
    else:
        downloaded_file_path = root / info.url.split("/")[-1]
```

Then, I think `anomalib`'s download function should work. You wouldn't need to overwrite it.

Let me know if it works. Cheers!

Reply:
Thank you for the great suggestion. I've implemented and tested the change and it works fine. It has been updated in the latest commits.